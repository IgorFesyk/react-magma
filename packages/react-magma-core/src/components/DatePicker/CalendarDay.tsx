import * as React from 'react';

export interface CalendarDayCoreProps {
  children: (props) => React.ReactNode;
  // onClick?: (event: React.SyntheticEvent) => void;
}

export class CalendarDayCore extends React.Component<CalendarDayCoreProps> {
  constructor(props) {
    super(props);

    // this.onClick = this.onClick.bind(this);
  }

  // onClick(event) {
  //   this.props.onClick && this.props.onClick(event);
  // }

  render() {
    return this.props.children({
      ...this.props
      // onClick: this.onClick
    });
  }
}
