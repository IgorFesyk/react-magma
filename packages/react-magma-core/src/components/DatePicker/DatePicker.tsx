import * as React from 'react';
import {
  getCalendarMonthWeeks,
  getPrevMonthFromDate,
  getNextMonthFromDate,
  handleKeyPress
} from './utils';

export interface DatePickerCoreProps {
  children: (props) => React.ReactNode;
  calendarOpened?: boolean;
  defaultDate?: Date;
  onDayClick?: (day: any, event: React.SyntheticEvent) => void;
}

interface DatePickerState {
  calendarOpened: boolean;
  focusedDate: Date;
  chosenDate?: Date;
  dateFocused?: boolean;
}

export class DatePickerCore extends React.Component<
  DatePickerCoreProps,
  DatePickerState
> {
  initialState: DatePickerState = {
    calendarOpened: this.props.calendarOpened,
    focusedDate: this.props.defaultDate || new Date(),
    chosenDate: this.props.defaultDate
  };
  state: DatePickerState = this.initialState;

  constructor(props) {
    super(props);

    this.onInputFocus = this.onInputFocus.bind(this);
    this.onDateFocus = this.onDateFocus.bind(this);
    this.onCalendarBlur = this.onCalendarBlur.bind(this);
    this.onHelperFocus = this.onHelperFocus.bind(this);
    this.onPrevMonthClick = this.onPrevMonthClick.bind(this);
    this.onNextMonthClick = this.onNextMonthClick.bind(this);
    this.onKeyDown = this.onKeyDown.bind(this);
    this.onEscKey = this.onEscKey.bind(this);
    this.onDayChangedByKeyboardNavigation = this.onDayChangedByKeyboardNavigation.bind(
      this
    );
    this.onDayClick = this.onDayClick.bind(this);
  }

  buildCalendarMonth(date: Date, enableOutsideDates: boolean) {
    return getCalendarMonthWeeks(date, enableOutsideDates);
  }

  onInputFocus() {
    this.setState({ calendarOpened: true });
  }

  onDateFocus() {
    this.setState({ dateFocused: true });
  }

  onHelperFocus() {
    this.setState({ dateFocused: false });
  }

  onCalendarBlur(event) {
    const { currentTarget } = event;

    // timeout needed for active element to update. Browser behavior.
    // https://bugzilla.mozilla.org/show_bug.cgi?id=452307
    setTimeout(() => {
      const isInStuff = currentTarget.contains(document.activeElement);

      if (!isInStuff) {
        this.setState({ calendarOpened: false });
      }
    }, 0);
  }

  onPrevMonthClick() {
    this.onDayChangedByKeyboardNavigation(
      getPrevMonthFromDate(this.state.focusedDate)
    );
  }

  onNextMonthClick() {
    this.onDayChangedByKeyboardNavigation(
      getNextMonthFromDate(this.state.focusedDate)
    );
  }

  onKeyDown(event: React.KeyboardEvent) {
    if (this.state.dateFocused) {
      const newChosenDate = handleKeyPress(
        event,
        this.state.focusedDate,
        this.onEscKey,
        this.onDayClick
      );
      if (newChosenDate) {
        this.onDayChangedByKeyboardNavigation(newChosenDate);
      }
    }
  }

  onEscKey() {
    this.setState({ calendarOpened: false });
  }

  onDayChangedByKeyboardNavigation(day) {
    this.setState({ focusedDate: day });
  }

  onDayClick(day, event) {
    this.props.onDayClick && this.props.onDayClick(day, event);

    this.setState({ chosenDate: day, calendarOpened: false });
  }

  render() {
    const { calendarOpened, chosenDate, focusedDate, dateFocused } = this.state;

    return this.props.children({
      ...this.props,
      calendarOpened,
      chosenDate,
      focusedDate,
      dateFocused,
      buildCalendarMonth: this.buildCalendarMonth,
      onCalendarBlur: this.onCalendarBlur,
      onInputFocus: this.onInputFocus,
      onDateFocus: this.onDateFocus,
      onHelperFocus: this.onHelperFocus,
      onPrevMonthClick: this.onPrevMonthClick,
      onNextMonthClick: this.onNextMonthClick,
      onKeyDown: this.onKeyDown,
      onDayClick: this.onDayClick
    });
  }
}
