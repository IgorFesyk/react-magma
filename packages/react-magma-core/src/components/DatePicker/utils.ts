import {
  subDays,
  subWeeks,
  subMonths,
  addDays,
  addWeeks,
  addMonths,
  getDay,
  startOfWeek,
  startOfMonth,
  endOfWeek,
  endOfMonth,
  differenceInDays
} from 'date-fns';

export function getCalendarMonthWeeks(
  month,
  enableOutsideDays,
  firstDayOfWeek = 0
) {
  const firstOfMonth = startOfMonth(month);
  const lastOfMonth = endOfMonth(month);

  const prevDays = (getDay(firstOfMonth) + 7 - firstDayOfWeek) % 7;
  const nextDays = (firstDayOfWeek + 6 - getDay(lastOfMonth)) % 7;
  const firstDay = subDays(firstOfMonth, prevDays);
  const lastDay = addDays(lastOfMonth, nextDays);

  const totalDays = differenceInDays(lastDay, firstDay) + 1;

  const currentDay = firstDay;
  const weeksInMonth = [];

  for (let i = 0; i < totalDays; i += 1) {
    if (i % 7 === 0) {
      weeksInMonth.push([]);
    }

    let day = null;
    if ((i >= prevDays && i < totalDays - nextDays) || enableOutsideDays) {
      day = addDays(currentDay, i);
    }

    weeksInMonth[weeksInMonth.length - 1].push(day);
  }

  return weeksInMonth;
}

export function getPrevMonthFromDate(prevDate) {
  return startOfMonth(subMonths(prevDate, 1));
}

export function getNextMonthFromDate(prevDate) {
  return startOfMonth(addMonths(prevDate, 1));
}

export function handleKeyPress(
  e: React.KeyboardEvent,
  prevDate: Date,
  onEscKey?: () => void,
  onDayClick?: (day: Date, event: React.SyntheticEvent) => void
) {
  const { key } = e;

  if (key === 'Enter' || key === ' ') {
    onDayClick(prevDate, e);
  }

  switch (key) {
    case 'ArrowUp':
      e.preventDefault();
      return subWeeks(prevDate, 1);

    case 'ArrowLeft':
      e.preventDefault();
      return subDays(prevDate, 1);

    case 'Home':
      e.preventDefault();
      return startOfWeek(prevDate);

    case 'PageUp':
      e.preventDefault();
      return subMonths(prevDate, 1);

    case 'ArrowDown':
      e.preventDefault();
      return addWeeks(prevDate, 1);

    case 'ArrowRight':
      e.preventDefault();
      return addDays(prevDate, 1);

    case 'End':
      e.preventDefault();
      return endOfWeek(prevDate);

    case 'PageDown':
      e.preventDefault();
      return addMonths(prevDate, 1);

    case 'Escape':
      e.preventDefault();
      onEscKey();
      break;

    default:
      break;
  }
}
