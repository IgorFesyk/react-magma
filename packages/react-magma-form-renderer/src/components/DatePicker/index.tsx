import React from 'react';
import useFieldApi from '@data-driven-forms/react-form-renderer/dist/cjs/use-field-api';
import { UseFieldApiConfig } from '@data-driven-forms/react-form-renderer';
import { v4 as uuidv4 } from 'uuid';
import {
  DatePicker as MagmaDatePicker,
  DatePickerProps as MagmaDatePickerProps,
} from 'react-magma-dom';

export type DatePickerProps = MagmaDatePickerProps & UseFieldApiConfig;

const DatePickerMapping: React.FunctionComponent<DatePickerProps> = props => {
  const {
    input,
    validateOnMount,
    showError,
    type,
    meta: { error, submitFailed },
    ...rest
  } = useFieldApi(props);
  const id = input.name || uuidv4();
  const errorMessage =
    ((validateOnMount || submitFailed || showError) && error) || '';

  return (
    <MagmaDatePicker
      {...input}
      id={id}
      errorMessage={errorMessage}
      labelText={rest.labelText}
      {...rest}
    />
  );
};

export const DatePicker = React.memo(DatePickerMapping);
