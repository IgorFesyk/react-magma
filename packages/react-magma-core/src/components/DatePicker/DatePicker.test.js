/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
import * as React from 'react';
// import { mount } from 'enzyme';
import { render, fireEvent } from 'react-testing-library';
import { DatePickerCore } from './DatePicker';
import {
  isSameDay,
  subMonths,
  addMonths,
  startOfMonth,
  addDays,
  subWeeks,
  subDays,
  startOfWeek,
  addWeeks,
  endOfWeek,
  format
} from 'date-fns';

const onDayClick = jest.fn();

const DATE_PICKER_CORE_PROPS = {
  children: () => React.createElement('div'),
  calendarOpened: false,
  defaultDate: new Date('January 17, 2019'),
  onDayClick
};

describe('DatePickerCore', () => {
  afterEach(() => {
    onDayClick.mockReset();
  });

  describe('state management', () => {
    it('should create the initial state of the input', () => {
      const defaultDate = new Date();
      const formattedDate = format(defaultDate, 'MMMM Do YYYY');
      const { getByTestId } = render(
        <DatePickerCore calendarOpened={true} defaultDate={defaultDate}>
          {({ calendarOpened, focusedDate, chosenDate }) => {
            return (
              <>
                <input
                  data-testid="calendarOpened"
                  type="checkbox"
                  checked={calendarOpened}
                />
                <span data-testid="focusedDate">
                  {format(focusedDate, 'MMMM Do YYYY')}
                </span>
                <span data-testid="chosenDate">
                  {format(chosenDate, 'MMMM Do YYYY')}
                </span>
              </>
            );
          }}
        </DatePickerCore>
      );

      expect(getByTestId('calendarOpened').checked).toEqual(true);
      expect(getByTestId('focusedDate')).toHaveTextContent(formattedDate);
      expect(getByTestId('chosenDate')).toHaveTextContent(formattedDate);
    });

    it('should default focused date to todays date if no default day is sent in', () => {
      const { getByTestId } = render(
        <DatePickerCore>
          {({ focusedDate }) => {
            return (
              <>
                <span data-testid="focusedDate" data-date={focusedDate} />
              </>
            );
          }}
        </DatePickerCore>
      );

      expect(
        isSameDay(
          getByTestId('focusedDate').getAttribute('data-date'),
          new Date()
        )
      ).toBeTruthy();
    });
  });

  describe('build calendar month', () => {
    it('should build the calendar for the given date', () => {
      const defaultDate = new Date('January 17, 2019');
      const { getByTestId } = render(
        <DatePickerCore>
          {({ buildCalendarMonth }) => {
            const calendarData = React.createRef();
            return (
              <>
                <button
                  data-testid="buildCalendarButton"
                  onClick={() => {
                    expect(buildCalendarMonth(defaultDate)).toMatchSnapshot();
                  }}
                >
                  Build Calendar
                </button>
                <span
                  data-testid="calendarData"
                  ref={calendarData}
                  data-calendarData=""
                />
              </>
            );
          }}
        </DatePickerCore>
      );

      fireEvent.click(getByTestId('buildCalendarButton'));
    });

    it('should build the calendar for the given date with outside dates enabled', () => {
      const defaultDate = new Date('January 17, 2019');
      const { getByTestId } = render(
        <DatePickerCore>
          {({ buildCalendarMonth }) => {
            return (
              <>
                <button
                  data-testid="buildCalendarButton"
                  onClick={() => {
                    expect(
                      buildCalendarMonth(defaultDate, true)
                    ).toMatchSnapshot();
                  }}
                >
                  Build Calendar
                </button>
              </>
            );
          }}
        </DatePickerCore>
      );

      fireEvent.click(getByTestId('buildCalendarButton'));
    });
  });

  describe('handle calendar focus', () => {
    it('should open the calendar on focus', () => {
      const { getByTestId } = render(
        <DatePickerCore>
          {({ calendarOpened, onInputFocus }) => {
            return (
              <>
                <input data-testid="inputToFocus" onFocus={onInputFocus} />
                <span
                  data-testid="calendarOpened"
                  data-calendarOpened={calendarOpened}
                />
              </>
            );
          }}
        </DatePickerCore>
      );

      expect(
        getByTestId('calendarOpened').getAttribute('data-calendarOpened')
      ).toBeFalsy();

      fireEvent.focus(getByTestId('inputToFocus'));

      expect(
        getByTestId('calendarOpened').getAttribute('data-calendarOpened')
      ).toBeTruthy();
    });
  });

  describe('Focus and blur', () => {
    beforeEach(() => {
      jest.useFakeTimers();
    });

    afterEach(() => {
      jest.useRealTimers();
    });

    it('should handle date focus', () => {
      const { getByTestId } = render(
        <DatePickerCore>
          {({ dateFocused, onDateFocus, onHelperFocus }) => {
            return (
              <>
                <div data-testid="focusDate" onFocus={onDateFocus} />
                <div data-testid="focusHelper" onFocus={onHelperFocus} />
                <input data-testid="dateFocused" checked={dateFocused} />
              </>
            );
          }}
        </DatePickerCore>
      );

      fireEvent.focus(getByTestId('focusDate'));

      expect(getByTestId('dateFocused').checked).toBeTruthy();

      fireEvent.focus(getByTestId('focusHelper'));

      expect(getByTestId('dateFocused').checked).toBeFalsy();
    });

    it('should close the calendar when active element is inside of calendar container', () => {
      const { container } = render(
        <DatePickerCore {...DATE_PICKER_CORE_PROPS} calendarOpened={true}>
          {({ onCalendarBlur, calendarOpened }) => {
            return (
              <>
                <div
                  style={{ display: calendarOpened ? 'block' : 'none' }}
                  id="calendar"
                  onBlur={onCalendarBlur}
                >
                  <button id="calendarButton" />
                </div>
                <button id="notInCalendarButton" />
              </>
            );
          }}
        </DatePickerCore>
      );

      const calendarButton = container.querySelector('#calendarButton');
      const notCalendarButton = container.querySelector('#notInCalendarButton');

      calendarButton.focus();
      notCalendarButton.focus();

      jest.advanceTimersByTime(1000);

      expect(calendarButton).not.toBeVisible();
    });

    it('should not close the calendar when active element is outside of calendar container', () => {
      const { container } = render(
        <DatePickerCore {...DATE_PICKER_CORE_PROPS} calendarOpened={true}>
          {({ onCalendarBlur, calendarOpened }) => {
            return (
              <>
                <div
                  style={{ display: calendarOpened ? 'block' : 'none' }}
                  id="calendar"
                  onBlur={onCalendarBlur}
                >
                  <button id="calendarButton" />
                  <button id="calendarButton2" />
                </div>
                <button id="notInCalendarButton" />
              </>
            );
          }}
        </DatePickerCore>
      );

      const calendarButton = container.querySelector('#calendarButton');
      const calendarButton2 = container.querySelector('#calendarButton2');

      calendarButton.focus();
      calendarButton2.focus();

      jest.advanceTimersByTime(1000);

      expect(calendarButton2).toBeVisible();
    });
  });

  describe('on previous button click', () => {
    it('changes the focus date to the first of the previous month', () => {
      const defaultDate = new Date();
      const { getByTestId } = render(
        <DatePickerCore defaultDate={defaultDate}>
          {({ focusedDate, onPrevMonthClick }) => {
            return (
              <>
                <button
                  data-testid="prevMonthButton"
                  onClick={onPrevMonthClick}
                >
                  Previous Month
                </button>
                <span data-testid="focusedDate" data-date={focusedDate} />
              </>
            );
          }}
        </DatePickerCore>
      );

      fireEvent.click(getByTestId('prevMonthButton'));

      expect(
        isSameDay(
          getByTestId('focusedDate').getAttribute('data-date'),
          startOfMonth(subMonths(defaultDate, 1))
        )
      ).toBeTruthy();
    });
  });

  describe('on next button click', () => {
    it('changes the focus date to the first of the next month', () => {
      const defaultDate = new Date();
      const { getByTestId } = render(
        <DatePickerCore defaultDate={defaultDate}>
          {({ focusedDate, onNextMonthClick }) => {
            return (
              <>
                <button
                  data-testid="nextMonthButton"
                  onClick={onNextMonthClick}
                >
                  Previous Month
                </button>
                <span data-testid="focusedDate" data-date={focusedDate} />
              </>
            );
          }}
        </DatePickerCore>
      );

      fireEvent.click(getByTestId('nextMonthButton'));

      expect(
        isSameDay(
          getByTestId('focusedDate').getAttribute('data-date'),
          startOfMonth(addMonths(defaultDate, 1))
        )
      ).toBeTruthy();
    });
  });

  describe('on key down press', () => {
    it('does not update focused date if date is not focused', () => {
      const defaultDate = new Date();
      const { getByTestId } = render(
        <DatePickerCore defaultDate={defaultDate}>
          {({ focusedDate, onKeyDown }) => {
            return (
              <>
                <div
                  data-testid="calendarContainer"
                  role="table"
                  onKeyDown={onKeyDown}
                >
                  <span data-testid="focusedDate" data-date={focusedDate} />
                </div>
              </>
            );
          }}
        </DatePickerCore>
      );

      fireEvent.keyDown(getByTestId('calendarContainer'), {
        key: 'ArrowUp',
        code: 38
      });

      expect(
        isSameDay(
          getByTestId('focusedDate').getAttribute('data-date'),
          defaultDate
        )
      ).toBeTruthy();
    });

    it('ArrowUp', () => {
      const defaultDate = new Date();
      const { getByTestId } = render(
        <DatePickerCore defaultDate={defaultDate}>
          {({ focusedDate, onKeyDown, onDateFocus }) => {
            return (
              <>
                <div
                  data-testid="calendarContainer"
                  role="table"
                  onFocus={onDateFocus}
                  onKeyDown={onKeyDown}
                >
                  <span data-testid="focusedDate" data-date={focusedDate} />
                </div>
              </>
            );
          }}
        </DatePickerCore>
      );

      fireEvent.focus(getByTestId('calendarContainer'));
      fireEvent.keyDown(getByTestId('calendarContainer'), {
        key: 'ArrowUp',
        code: 38
      });

      expect(
        isSameDay(
          getByTestId('focusedDate').getAttribute('data-date'),
          subWeeks(defaultDate, 1)
        )
      ).toBeTruthy();
    });

    it('ArrowLeft', () => {
      const defaultDate = new Date();
      const { getByTestId } = render(
        <DatePickerCore defaultDate={defaultDate}>
          {({ focusedDate, onKeyDown, onDateFocus }) => {
            return (
              <>
                <div
                  data-testid="calendarContainer"
                  role="table"
                  onFocus={onDateFocus}
                  onKeyDown={onKeyDown}
                >
                  <span data-testid="focusedDate" data-date={focusedDate} />
                </div>
              </>
            );
          }}
        </DatePickerCore>
      );

      fireEvent.focus(getByTestId('calendarContainer'));
      fireEvent.keyDown(getByTestId('calendarContainer'), {
        key: 'ArrowLeft',
        code: 37
      });

      expect(
        isSameDay(
          getByTestId('focusedDate').getAttribute('data-date'),
          subDays(defaultDate, 1)
        )
      ).toBeTruthy();
    });

    it('Home', () => {
      const defaultDate = new Date();
      const { getByTestId } = render(
        <DatePickerCore defaultDate={defaultDate}>
          {({ focusedDate, onKeyDown, onDateFocus }) => {
            return (
              <>
                <div
                  data-testid="calendarContainer"
                  role="table"
                  onFocus={onDateFocus}
                  onKeyDown={onKeyDown}
                >
                  <span data-testid="focusedDate" data-date={focusedDate} />
                </div>
              </>
            );
          }}
        </DatePickerCore>
      );

      fireEvent.focus(getByTestId('calendarContainer'));
      fireEvent.keyDown(getByTestId('calendarContainer'), {
        key: 'Home',
        code: 36
      });

      expect(
        isSameDay(
          getByTestId('focusedDate').getAttribute('data-date'),
          startOfWeek(defaultDate)
        )
      ).toBeTruthy();
    });

    it('PageUp', () => {
      const defaultDate = new Date();
      const { getByTestId } = render(
        <DatePickerCore defaultDate={defaultDate}>
          {({ focusedDate, onKeyDown, onDateFocus }) => {
            return (
              <>
                <div
                  data-testid="calendarContainer"
                  role="table"
                  onFocus={onDateFocus}
                  onKeyDown={onKeyDown}
                >
                  <span data-testid="focusedDate" data-date={focusedDate} />
                </div>
              </>
            );
          }}
        </DatePickerCore>
      );

      fireEvent.focus(getByTestId('calendarContainer'));
      fireEvent.keyDown(getByTestId('calendarContainer'), {
        key: 'PageUp',
        code: 33
      });

      expect(
        isSameDay(
          getByTestId('focusedDate').getAttribute('data-date'),
          subMonths(defaultDate, 1)
        )
      ).toBeTruthy();
    });

    it('PageDown', () => {
      const defaultDate = new Date();
      const { getByTestId } = render(
        <DatePickerCore defaultDate={defaultDate}>
          {({ focusedDate, onKeyDown, onDateFocus }) => {
            return (
              <>
                <div
                  data-testid="calendarContainer"
                  role="table"
                  onFocus={onDateFocus}
                  onKeyDown={onKeyDown}
                >
                  <span data-testid="focusedDate" data-date={focusedDate} />
                </div>
              </>
            );
          }}
        </DatePickerCore>
      );

      fireEvent.focus(getByTestId('calendarContainer'));
      fireEvent.keyDown(getByTestId('calendarContainer'), {
        key: 'PageDown',
        code: 34
      });

      expect(
        isSameDay(
          getByTestId('focusedDate').getAttribute('data-date'),
          addMonths(defaultDate, 1)
        )
      ).toBeTruthy();
    });

    it('ArrowDown', () => {
      const defaultDate = new Date();
      const { getByTestId } = render(
        <DatePickerCore defaultDate={defaultDate}>
          {({ focusedDate, onKeyDown, onDateFocus }) => {
            return (
              <>
                <div
                  data-testid="calendarContainer"
                  role="table"
                  onFocus={onDateFocus}
                  onKeyDown={onKeyDown}
                >
                  <span data-testid="focusedDate" data-date={focusedDate} />
                </div>
              </>
            );
          }}
        </DatePickerCore>
      );

      fireEvent.focus(getByTestId('calendarContainer'));
      fireEvent.keyDown(getByTestId('calendarContainer'), {
        key: 'ArrowDown',
        code: 40
      });

      expect(
        isSameDay(
          getByTestId('focusedDate').getAttribute('data-date'),
          addWeeks(defaultDate, 1)
        )
      ).toBeTruthy();
    });

    it('ArrowRight', () => {
      const defaultDate = new Date();
      const { getByTestId } = render(
        <DatePickerCore defaultDate={defaultDate}>
          {({ focusedDate, onKeyDown, onDateFocus }) => {
            return (
              <>
                <div
                  data-testid="calendarContainer"
                  role="table"
                  onFocus={onDateFocus}
                  onKeyDown={onKeyDown}
                >
                  <span data-testid="focusedDate" data-date={focusedDate} />
                </div>
              </>
            );
          }}
        </DatePickerCore>
      );

      fireEvent.focus(getByTestId('calendarContainer'));
      fireEvent.keyDown(getByTestId('calendarContainer'), {
        key: 'ArrowRight',
        code: 39
      });

      expect(
        isSameDay(
          getByTestId('focusedDate').getAttribute('data-date'),
          addDays(defaultDate, 1)
        )
      ).toBeTruthy();
    });

    it('End', () => {
      const defaultDate = new Date();
      const { getByTestId } = render(
        <DatePickerCore defaultDate={defaultDate}>
          {({ focusedDate, onKeyDown, onDateFocus }) => {
            return (
              <>
                <div
                  data-testid="calendarContainer"
                  role="table"
                  onFocus={onDateFocus}
                  onKeyDown={onKeyDown}
                >
                  <span data-testid="focusedDate" data-date={focusedDate} />
                </div>
              </>
            );
          }}
        </DatePickerCore>
      );

      fireEvent.focus(getByTestId('calendarContainer'));
      fireEvent.keyDown(getByTestId('calendarContainer'), {
        key: 'End',
        code: 35
      });

      expect(
        isSameDay(
          getByTestId('focusedDate').getAttribute('data-date'),
          endOfWeek(defaultDate)
        )
      ).toBeTruthy();
    });

    it('Escape', () => {
      const defaultDate = new Date();
      const { getByTestId } = render(
        <DatePickerCore calendarOpened={true} defaultDate={defaultDate}>
          {({ calendarOpened, onKeyDown, onDateFocus }) => {
            return (
              <>
                <div
                  data-testid="calendarContainer"
                  role="table"
                  onFocus={onDateFocus}
                  onKeyDown={onKeyDown}
                >
                  <input
                    data-testid="calendarOpened"
                    checked={calendarOpened}
                  />
                </div>
              </>
            );
          }}
        </DatePickerCore>
      );

      fireEvent.focus(getByTestId('calendarContainer'));
      fireEvent.keyDown(getByTestId('calendarContainer'), {
        key: 'Escape',
        code: 27
      });

      expect(getByTestId('calendarOpened').checked).toBeFalsy();
    });

    it('Enter', () => {
      const defaultDate = new Date();
      const { getByTestId } = render(
        <DatePickerCore defaultDate={defaultDate}>
          {({ chosenDate, onKeyDown, onDateFocus }) => {
            return (
              <>
                <div
                  data-testid="calendarContainer"
                  role="table"
                  onFocus={onDateFocus}
                  onKeyDown={onKeyDown}
                >
                  <span data-testid="chosenDate" data-date={chosenDate} />
                </div>
              </>
            );
          }}
        </DatePickerCore>
      );

      fireEvent.focus(getByTestId('calendarContainer'));
      fireEvent.keyDown(getByTestId('calendarContainer'), {
        key: 'Enter',
        code: 13
      });

      expect(
        isSameDay(
          getByTestId('chosenDate').getAttribute('data-date'),
          defaultDate
        )
      ).toBeTruthy();
    });

    it('Spacebar', () => {
      const defaultDate = new Date();
      const { getByTestId } = render(
        <DatePickerCore defaultDate={defaultDate}>
          {({ chosenDate, onKeyDown, onDateFocus }) => {
            return (
              <>
                <div
                  data-testid="calendarContainer"
                  role="table"
                  onFocus={onDateFocus}
                  onKeyDown={onKeyDown}
                >
                  <span data-testid="chosenDate" data-date={chosenDate} />
                </div>
              </>
            );
          }}
        </DatePickerCore>
      );

      fireEvent.focus(getByTestId('calendarContainer'));
      fireEvent.keyDown(getByTestId('calendarContainer'), {
        key: 'Space',
        code: 32
      });

      expect(
        isSameDay(
          getByTestId('chosenDate').getAttribute('data-date'),
          defaultDate
        )
      ).toBeTruthy();
    });

    it('does not update the focused date if a bad key press occurs', () => {
      const defaultDate = new Date();
      const { getByTestId } = render(
        <DatePickerCore defaultDate={defaultDate}>
          {({ focusedDate, onKeyDown, onDateFocus }) => {
            return (
              <>
                <div
                  data-testid="calendarContainer"
                  role="table"
                  onFocus={onDateFocus}
                  onKeyDown={onKeyDown}
                >
                  <span data-testid="focusedDate" data-date={focusedDate} />
                </div>
              </>
            );
          }}
        </DatePickerCore>
      );

      fireEvent.focus(getByTestId('calendarContainer'));
      fireEvent.keyDown(getByTestId('calendarContainer'), {
        key: 'f',
        code: 70
      });

      expect(
        isSameDay(
          getByTestId('focusedDate').getAttribute('data-date'),
          defaultDate
        )
      ).toBeTruthy();
    });
  });

  // describe('on esc key', () => {
  //   it('closes the calendar when the esc key function is called', () => {
  //     const component = datePickerSetup({ calendarOpened: true });
  //     component.setState({ dateFocused: true });

  //     component.instance().onEscKey();

  //     expect(component.state('calendarOpened')).toBeFalsy();
  //   });
  // });

  // describe('on day changed by keyboard navigation', () => {
  //   it('updates the focused date when the keyboard naviagation function is called', () => {
  //     const updatedDate = new Date();
  //     const component = datePickerSetup({ calendarOpened: true });
  //     component.setState({ dateFocused: true });

  //     component.instance().onDayChangedByKeyboardNavigation(updatedDate);

  //     expect(component.state('focusedDate')).toEqual(updatedDate);
  //   });
  // });

  describe('on day click', () => {
    it('should updated the chosen date and close the calendar', () => {
      const defaultDate = new Date();
      const newDate = addDays(defaultDate, 1);
      const { getByTestId } = render(
        <DatePickerCore calendarOpened={true} defaultDate={defaultDate}>
          {({ calendarOpened, chosenDate, onDateFocus, onDayClick }) => {
            return (
              <>
                <div
                  data-testid="calendarContainer"
                  role="table"
                  onFocus={onDateFocus}
                >
                  <button
                    data-testid="dayClickButton"
                    onClick={() => {
                      onDayClick(newDate);
                    }}
                  >
                    Click Day
                  </button>
                  <span data-testid="chosenDate" data-date={chosenDate} />
                  <input
                    data-testid="calendarOpened"
                    checked={calendarOpened}
                  />
                </div>
              </>
            );
          }}
        </DatePickerCore>
      );

      fireEvent.focus(getByTestId('calendarContainer'));
      fireEvent.click(getByTestId('dayClickButton'));

      expect(
        isSameDay(getByTestId('chosenDate').getAttribute('data-date'), newDate)
      ).toBeTruthy();
      expect(getByTestId('calendarOpened').checked).toBeFalsy();
    });

    it('should call the passed in onDayClick function', () => {
      const onDayClickSpy = jest.fn();
      const defaultDate = new Date();
      const newDate = addDays(defaultDate, 1);
      const { getByTestId } = render(
        <DatePickerCore
          calendarOpened={true}
          defaultDate={defaultDate}
          onDayClick={onDayClickSpy}
        >
          {({ onDayClick }) => {
            return (
              <>
                <div data-testid="calendarContainer" role="table">
                  <button
                    data-testid="dayClickButton"
                    onClick={() => {
                      onDayClick(newDate);
                    }}
                  >
                    Click Day
                  </button>
                </div>
              </>
            );
          }}
        </DatePickerCore>
      );

      const event = {};
      fireEvent.focus(getByTestId('calendarContainer'));
      fireEvent.click(getByTestId('dayClickButton'), event);

      expect(onDayClickSpy).toHaveBeenCalled();
    });
  });
});
