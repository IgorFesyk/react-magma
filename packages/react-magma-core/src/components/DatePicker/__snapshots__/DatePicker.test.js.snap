// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePickerCore build calendar month should build the calendar for the given date 1`] = `
Array [
  Array [
    null,
    null,
    2019-01-01T05:00:00.000Z,
    2019-01-02T05:00:00.000Z,
    2019-01-03T05:00:00.000Z,
    2019-01-04T05:00:00.000Z,
    2019-01-05T05:00:00.000Z,
  ],
  Array [
    2019-01-06T05:00:00.000Z,
    2019-01-07T05:00:00.000Z,
    2019-01-08T05:00:00.000Z,
    2019-01-09T05:00:00.000Z,
    2019-01-10T05:00:00.000Z,
    2019-01-11T05:00:00.000Z,
    2019-01-12T05:00:00.000Z,
  ],
  Array [
    2019-01-13T05:00:00.000Z,
    2019-01-14T05:00:00.000Z,
    2019-01-15T05:00:00.000Z,
    2019-01-16T05:00:00.000Z,
    2019-01-17T05:00:00.000Z,
    2019-01-18T05:00:00.000Z,
    2019-01-19T05:00:00.000Z,
  ],
  Array [
    2019-01-20T05:00:00.000Z,
    2019-01-21T05:00:00.000Z,
    2019-01-22T05:00:00.000Z,
    2019-01-23T05:00:00.000Z,
    2019-01-24T05:00:00.000Z,
    2019-01-25T05:00:00.000Z,
    2019-01-26T05:00:00.000Z,
  ],
  Array [
    2019-01-27T05:00:00.000Z,
    2019-01-28T05:00:00.000Z,
    2019-01-29T05:00:00.000Z,
    2019-01-30T05:00:00.000Z,
    2019-01-31T05:00:00.000Z,
    null,
    null,
  ],
]
`;

exports[`DatePickerCore build calendar month should build the calendar for the given date with outside dates enabled 1`] = `
Array [
  Array [
    2018-12-30T05:00:00.000Z,
    2018-12-31T05:00:00.000Z,
    2019-01-01T05:00:00.000Z,
    2019-01-02T05:00:00.000Z,
    2019-01-03T05:00:00.000Z,
    2019-01-04T05:00:00.000Z,
    2019-01-05T05:00:00.000Z,
  ],
  Array [
    2019-01-06T05:00:00.000Z,
    2019-01-07T05:00:00.000Z,
    2019-01-08T05:00:00.000Z,
    2019-01-09T05:00:00.000Z,
    2019-01-10T05:00:00.000Z,
    2019-01-11T05:00:00.000Z,
    2019-01-12T05:00:00.000Z,
  ],
  Array [
    2019-01-13T05:00:00.000Z,
    2019-01-14T05:00:00.000Z,
    2019-01-15T05:00:00.000Z,
    2019-01-16T05:00:00.000Z,
    2019-01-17T05:00:00.000Z,
    2019-01-18T05:00:00.000Z,
    2019-01-19T05:00:00.000Z,
  ],
  Array [
    2019-01-20T05:00:00.000Z,
    2019-01-21T05:00:00.000Z,
    2019-01-22T05:00:00.000Z,
    2019-01-23T05:00:00.000Z,
    2019-01-24T05:00:00.000Z,
    2019-01-25T05:00:00.000Z,
    2019-01-26T05:00:00.000Z,
  ],
  Array [
    2019-01-27T05:00:00.000Z,
    2019-01-28T05:00:00.000Z,
    2019-01-29T05:00:00.000Z,
    2019-01-30T05:00:00.000Z,
    2019-01-31T05:00:00.000Z,
    2019-02-01T05:00:00.000Z,
    2019-02-02T05:00:00.000Z,
  ],
]
`;
